contracts/optimism/TokenRateOracle.sol

```diff
@@ -240,31 +326,55 @@ contract TokenRateOracle is CrossDomainEnabled, ITokenRateOracle, Versioned {
         return false;
     }
 
-    function _setTokenRateAndTimestamps(uint128 tokenRate_, uint64 rateUpdateL1Timestamp_) internal {
-        _loadTokenRateData().value = TokenRateData(tokenRate_, rateUpdateL1Timestamp_, uint64(block.timestamp));
+    function _addTokenRate(
+        uint128 tokenRate_, uint64 rateUpdateL1Timestamp_, uint64 rateReceivedL2Timestamp_
+    ) internal {
+        _getStorageTokenRates().push(TokenRateData({
+            tokenRate: tokenRate_,
+            rateUpdateL1Timestamp: rateUpdateL1Timestamp_,
+            rateReceivedL2Timestamp: rateReceivedL2Timestamp_
+        }));
     }
 
-    function _getTokenRateAndTimestamps() private view returns (TokenRateData memory tokenRateData) {
-        tokenRateData = _loadTokenRateData().value;
+    function _getLastTokenRate() internal view returns (TokenRateData storage) {
+        if (_getStorageTokenRates().length == 0) {
+            revert ErrorNoTokenRateUpdates();
+        }
+        return _getTokenRateByIndex(_getStorageTokenRates().length - 1);
     }
 
-    /// @dev Storage helper for TokenRateData
-    struct StorageTokenRateData {
-        TokenRateData value;
+    function _getTokenRateByIndex(uint256 tokenRateIndex_) internal view returns (TokenRateData storage) {
+        if (tokenRateIndex_ >= _getStorageTokenRates().length) {
+            revert ErrorWrongTokenRateIndex();
+        }
+        return _getStorageTokenRates()[tokenRateIndex_];
     }
 
-    /// @dev Returns the reference to the slot with TokenRateData struct
-    function _loadTokenRateData()
-        private
-        pure
-        returns (StorageTokenRateData storage r)
-    {
-        bytes32 slot = TOKEN_RATE_DATA_SLOT;
+    function _getStorageTokenRates() internal pure returns (TokenRateData [] storage result) {
+        bytes32 position = TOKEN_RATES_DATA_SLOT;
         assembly {
-            r.slot := slot
+            result.slot := position
         }
     }
 
+    function _removeElementsAfterIndex(uint256 tokenRateIndex_) internal {
+        uint256 tokenRatesLength = _getStorageTokenRates().length;
+        if (tokenRateIndex_ < tokenRatesLength) {
+            uint256 numberOfElementsToRemove = tokenRatesLength - tokenRateIndex_ - 1;
+            for (uint256 i = 0; i < numberOfElementsToRemove; i++) {
+                _getStorageTokenRates().pop();
+            }
```

Comment:
:warning:  It can become an unexecutable piece of code (imagine someone maliciously put 10'000 storage elements within a single day), that could be solved by:
- disallowing pushing too close L1 timestamps (let's say they should be sparse with a delta of 1 hour at least)
- have an extra cursor pointing most of the times to the last element of the array, but getting rewind back here

Also, we can't use 'shrink' tricks here https://github.com/OpenZeppelin/openzeppelin-contracts/issues/3491 because once shrunked, the array wouldn't be able to extend any further

Reply:
Very nice catch!
I've got a few amendments.
Disallowing pushing close L1 timestamps imposes assumption on L1 rate update timings, it is okay, but better to avoid if we can. Additionally, L1 rate update might quite happen at the end of frame (n-1) and at start of frame (n).
How about
1) not pushing rate if its value and L1 timestamp equal to the last element of the array, but only update L2 update timestamp?
2) add restriction on the update frequency based on L2 (not L1) timestamp? Make the delta less than 1 hour, e.g. 10 min. 10 min corresponds to max 144 new array elements per day. I've tried popping 300 elements it costed `1.3m gas for 2 days` - quite bearable.
3) if we dare to introduce a bit of assembly https://ethereum.stackexchange.com/a/51897 cost of popping (n) elements will be decreased to O(1). Or, as you say, introduce cursor pointing to the last element, but afaiu it will require fully redo  array management from standard solidity to manual management. Otherwise the cursor will point to some inner element but array length will be inconsistent thought the code.

Reply:
Assembly solution might not work for us long-term, better opt-out from the item 3 as for me :pray:

Reply:
We agreed to reduce the frequency of token rate updates by not allowing them to be sent too often. This approach uses a configurable parameter, `MAX_ALLOWED_TIME_BETWEEEN_TOKEN_RATE_UPDATES`.